### PR TITLE
Implement kubernetes secret handling for podman play kube

### LINKED ISSUE #12396: podman play kube and secrets
I need some guidance. I can start pods successfully using "podman play kube" with podman secrets, provided to the container with fromEnv and secretRef. My problem with this solution is that then my secrets are viewable using inspect. Is it possible to have them mounted as files like with "podman run --secret"? Or is that just not possible right now with "podman play kube"?

### PR DESCRIPTION
**Pull request by cdoern**
add support for both creating a secret using yaml and mounting a secret as a volume given a yaml file.
Kubernetes secrets have a different structure than podman and therefore have to be handeled differently.
In this PR, I have introduced the basic usecases of kube secrets with more implementations like env secrets to come!

resolves #12396

Signed-off-by: cdoern <cdoern@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
podman now supports kubernetes secrets in podman play kube
```

### CODE REVIEW

**Comment by cdoern** [comment]
@vrothberg @ashley-cui @rhatdan @umohnani8 PTAL

File: pkg/domain/infra/abi/play.go
Comment (ashley-cui) [inline pkg/domain/infra/abi/play.go:994]: I'm concerned that we're only storing secret data, since later on we might want to support different types of kube secrets, and might want to store other kube specific metadata that exists in the original yaml. But I'm open to discussion on this

Comment (vrothberg) [reply pkg/domain/infra/abi/play.go:994]: I concur with @ashley-cui.

@cdoern can you elaborate on the unmarshal error?

File: pkg/domain/infra/abi/play.go
Comment (vrothberg) [inline pkg/domain/infra/abi/play.go:1002]: I imagine the `name` of the secret to be the _hash_/_digest_ of the secret struct.  The risk I see with using `secret.Name` is that it's not unique. By the laws of copy-pasting, I imagine many K8s yamls to use the same secret name.

Comment (vrothberg) [reply pkg/domain/infra/abi/play.go:1002]: We're doing something similar with the "service" containers of a K8s yaml (see https://github.com/containers/podman/blob/main/pkg/domain/infra/abi/play.go#L89).

File: pkg/domain/infra/abi/play.go
Comment (vrothberg) [inline pkg/domain/infra/abi/play.go:999]: `secretsPath` isn't used AFAIKS

File: test/e2e/play_kube_test.go
Comment (vrothberg) [inline test/e2e/play_kube_test.go:3890]: Nit: trailing empty line

File: test/e2e/play_kube_test.go
Comment (vrothberg) [inline test/e2e/play_kube_test.go:40]: I'd like an additional, more complex YAML with multiple secrets to make sure that mounting works for all secrets.

File: test/e2e/play_kube_test.go
Comment (vrothberg) [inline test/e2e/play_kube_test.go:3870]: Are secrets always in a different file or can they be specified alongside the pod/deployment?

**Review by vrothberg — COMMENTED** [review]
Really nice work, @cdoern.

That's a huge leap forward. Now we can iron out the details.

File: pkg/domain/infra/abi/play.go
Comment (cdoern) [reply pkg/domain/infra/abi/play.go:994]: so, in https://github.com/containers/podman/blob/0b630cd765d7ea1940b3962d352cf8d0826e65ca/pkg/specgen/generate/kube/volume.go#L137-L143 if you store the entire secret (marshaled) and thne unmarshal it into a v1.secret rather than a data map like I am here, the data in a v1.secret is a map[string][]byte but `yaml.unmarshal` does not understand that if you have "name: 1234" the 1234 can be unmarshaled as bytes rather than strings so we get a type mismatch error.

File: pkg/domain/infra/abi/play.go
Comment (cdoern) [reply pkg/domain/infra/abi/play.go:1002]: good point, I will change this around to use a hashing function

File: test/e2e/play_kube_test.go
Comment (cdoern) [reply test/e2e/play_kube_test.go:3870]: I haven't tested with a single file, I am not sure of the syntax of including both the pods and the secret within the same yaml, can this be done? All of the examples I saw online specified pre defined secrets in separate yaml files.

File: pkg/domain/infra/abi/play.go
Comment (cdoern) [reply pkg/domain/infra/abi/play.go:994]: I could make my own struct which would be almost identical but `Data` and `StringData` would be the same

Comment (ashley-cui) [reply pkg/domain/infra/abi/play.go:994]: Does json marshaling work with the original struct? If so, that could also be an option

Comment (cdoern) [reply pkg/domain/infra/abi/play.go:994]: @ashley-cui that does not seem to work either, only the custom struct seems to be possible here. I think that is a good option, right?

File: test/e2e/play_kube_test.go
Comment (cdoern) [reply test/e2e/play_kube_test.go:3870]: I see, you are looking for `---` examples? I think those will work, I will add a test using the `generateMultiDocKubeYaml` testing function

File: pkg/domain/infra/abi/play.go
Comment (vrothberg) [reply pkg/domain/infra/abi/play.go]: Sounds good to me.

File: pkg/domain/infra/abi/play.go
Comment (vrothberg) [reply pkg/domain/infra/abi/play.go:1005]: I think they are two separate things.  The hash would just be used internally by Podman to store and lookup the secret(s).

**Comment by cdoern** [comment]
@ashley-cui @vrothberg what do you think about the name issue?

**Comment by cdoern** [comment]
I think the best approach here is leaving the secrets with unique names. Either way, this change will need to be made in c/common not podman

**Comment by ashley-cui** [comment]
I'd tend to agree, as Kubernetnes allows separate files for secrets and the actual kube yaml, as well as a combined file. If we were to name the secret with a hash of the combined file, we would not be able to support separate files.

File: pkg/specgen/generate/kube/volume.go
Comment (vrothberg) [inline pkg/specgen/generate/kube/volume.go:14]: Can we use `.v3` instead?

File: go.mod
Comment (vrothberg) [inline go.mod:80]: This should probably be in the upper `require`.

File: pkg/domain/infra/abi/play.go
Comment (vrothberg) [inline pkg/domain/infra/abi/play.go:89]: The comment above needs updating.

**Review by vrothberg — COMMENTED** [review]
Nice work, @cdoern!  I agree on the naming issue.  Just a couple of nits.

Should we defer merging after we branched for 4.2?  I think/feel we should let it cook a bit and release with 4.3.

File: pkg/domain/infra/abi/play.go
Comment (vrothberg) [inline pkg/domain/infra/abi/play.go:1005]: I thought we'd use `k8sName(data)` as the name?

Comment (cdoern) [reply pkg/domain/infra/abi/play.go:1005]: @vrothberg lets talk at watercooler, I think using `k8sName` breaks the separate file usage of secrets since the user will not know the name given. I can change the function back to just being `serviceContainerName` but thought the change made the function more applicable to other types in the future.

**Review by flouthoc — CHANGES_REQUESTED** [review]
Since this was discussed before here are my two cents but please correct me if i made any errors.

* Secret name should be used as identifier from UX perspective.
* Creating secret with the same name should override the older secret. ( reason: one or more pod/deployment can use the same secret ) or error out with collision ( not preferred i don't think k8s does this )

* Add support for field `immutable` for erroring out on collision, so if `immutable: true` is set users cannot override or amend secret they must delete and recreate secret: https://kubernetes.io/docs/concepts/configuration/secret/#secret-immutable-create

File: pkg/domain/infra/abi/play.go
Comment (flouthoc) [reply pkg/domain/infra/abi/play.go:1005]: Added a comment below i think `secret.Name` is fine and on collision secret must be updated but if support for `immutable` is added then on collision it should error out for specific secrets where `immutable: true` is set. See below comment for more details.

Comment (vrothberg) [reply pkg/domain/infra/abi/play.go:1005]: Great input, @flouthoc! Thank you

**Comment by cdoern** [comment]
@flouthoc thanks for the review! @ashley-cui @vrothberg WDYT about replacing a secret when using the same name? I could wire that up in c/common.

**Comment by vrothberg** [comment]
I concur with Ashley.  Nothing IMHO should be changed in c/common.

**Comment by cdoern** [comment]
@ashley-cui @vrothberg as far as the immutable option that was brought up, that would need to be wired into c/common somehow right? or maybe we could write the metadata?

File: pkg/domain/infra/abi/play.go
Comment (vrothberg) [reply pkg/domain/infra/abi/play.go:1008]: Using metadata SGTM :+1:

**Comment by rhatdan** [comment]
@cdoern @ashley-cui @vrothberg is this ready for podman 4.2?

**Comment by cdoern** [comment]
@rhatdan, @vrothberg thinks it is best if we include this in 4.3

**Comment by rhatdan** [comment]
Ok we can wait til then

**Comment by vrothberg** [comment]
Since we branched for 4.2 already, we are not blocked.

**Comment by rhatdan** [comment]
/approved
@vrothberg Does this LGTY? @ashley-cui ?

File: pkg/domain/infra/abi/play.go
Comment (ashley-cui) [inline pkg/domain/infra/abi/play.go:993]: Does this comment still hold true? it looks like you're marshaling the entire *v1.Secret, not just the data

**Review by ashley-cui — COMMENTED** [review]
One small comment on a comment, but otherwise LGTM

**Comment by cdoern** [comment]
@ashley-cui @vrothberg should I get metadata handling into common before merging this? probably right?

**Comment by cdoern** [comment]
/hold 
until I get a common Pr and rebase this, it has been sitting about a week

**Comment by cdoern** [comment]
/hold cancel
@vrothberg @ashley-cui PTAL, vendored in c/common with my new changes.

**Comment by rhatdan** [comment]
Which PR are we holding up on containers/common?

**Comment by cdoern** [comment]
@rhatdan, it already merged. it was the metadata PR containers/common#1096

**Comment by cdoern** [comment]
@vrothberg @ashley-cui good to go here?

**Comment by vrothberg** [comment]
I will do a last review tomorrow morning. My brain is already fried for today ):

**Review by vrothberg — COMMENTED** [review]
/lgtm

**Comment by cdoern** [comment]
@flouthoc this is waiting for your approval I believe since you requested changes.

File: pkg/domain/infra/abi/play.go
Comment (flouthoc) [inline pkg/domain/infra/abi/play.go:1008]:
```suggestion
				return nil, fmt.Errorf("cannot modify existing secret as it is set to immutable")
```
Shouldn't this be modify ? I think at this point user is expecting to secret get created or get modified.

File: pkg/domain/infra/abi/play.go
Comment (flouthoc) [inline pkg/domain/infra/abi/play.go:1006]: Nit, non-blocking: string "immutable" is used couple of times across the code so i think can be a constant. But not worth a repush for now maybe in a different PR if this constant need to get declared in `c/common`.

File: pkg/domain/infra/abi/play.go
Comment (flouthoc) [inline pkg/domain/infra/abi/play.go:595]: Nit: extra line ( not worth a repush )

**Comment by flouthoc** [comment]
> @flouthoc this is waiting for your approval I believe since you requested changes.

@cdoern LGTM just a question about this https://github.com/containers/podman/pull/14567#discussion_r927597976 other comments are non-blocking.

File: pkg/domain/infra/abi/play.go
Comment (flouthoc) [reply pkg/domain/infra/abi/play.go:1006]: Resolving as this would probably need `c/common`.

File: pkg/domain/infra/abi/play.go
Comment (cdoern) [reply pkg/domain/infra/abi/play.go:1008]: I think `remove colliding` makes more sense since that's what is happening here on a podman level. We remove the colliding secret if we can ahd recreate a new one with possibly completely different config.

Comment (flouthoc) [reply pkg/domain/infra/abi/play.go:1008]: @cdoern Actually afaik k8s gives a error msg something on the lines of "unable to modify immutable field" but this is not worth blocking the PR. Could you leave this comment unresolved so I can confirm how it looks on k8s and  will update here and open a PR if need to.

**Review by flouthoc — APPROVED** [review]
LGTM
/lgtm
/approve

**Pull request merged**
